### PR TITLE
Fabo/776 fallback to mocked connection

### LINKED ISSUE #804: Check node version against gaiacli version
UI Version: 0.6.2

Description:
The node we are connecting to might have another version of the SDK then we are using and therefor the LCD might produce unexpected results. We should check the version and display an error if the node is on another version.

### LINKED ISSUE #851: Close menu on route change
UI Version: 0.7.0

Description:

If I click on the network indicator -> change network the menu is still open and I need to manually close it.

### PR DESCRIPTION
**Pull request by faboweb**
Closes #804
Closes #851

Description:

Do round robin on all known nodes. If there are no available nodes, show an error and let the user decide to either try all nodes again or switch to a mocked connection.

image

Also has code for #804 already

❤️ Thank you!

### CODE REVIEW

**Comment by faboweb** [comment]
Wait for SDK 0.20.0

File: app/src/main/index.js [hunk 1]
```diff
@@ -500,11 +500,14 @@ function pickNode(nodes) {
 }
 
 async function checkNodeSDKVersion() {
-  const nodeSDKVersion = await axios(
-    `http://localhost:${LCD_PORT}/version`
-  ).then(res => res.data)
+  const nodeSDKVersion = await axios
+    .get(`http://localhost:${LCD_PORT}/node_version`)
+    .then(res => res.data)
```
Comment (NodeGuy) [inline app/src/main/index.js]:
```JavaScript
(await axios.get(
    `http://localhost:${LCD_PORT}/node_version`
  )).data
```

File: test/unit/specs/main.spec.js [hunk 2]
```diff
@@ -367,9 +368,9 @@ describe("Startup Process", () => {
 
       registerIPCListeners(registeredIPCListeners)
 
-      // axios is used to ping nodes for the reconnection intent
+      // axios is used to ping nodes and get the SDK version of the node
       let axios = require("axios")
-      axios.get = () => Promise.resolve()
+      axios.get = () => Promise.resolve({ data: "0.19.0" })
```
Comment (NodeGuy) [inline test/unit/specs/main.spec.js]:
```JavaScript
axios.get = async () => ({ data: "0.19.0" })
```

File: test/unit/specs/main.spec.js [hunk 3]
```diff
@@ -693,6 +695,11 @@ function resetModulesKeepingFS() {
   // we want to keep Raven quiet
   const Raven = require("raven")
   Raven.disableConsoleAlerts()
+
+  // axios is used to ping nodes and get the SDK version of the node
+  // we need to keep this mocked, if not, the process fails
+  let axios = require("axios")
+  axios.get = () => Promise.resolve({ data: "0.19.0" })
```
Comment (NodeGuy) [inline test/unit/specs/main.spec.js]:
```JavaScript
axios.get = async () => ({ data: "0.19.0" })
```

**Comment by jolesbi** [comment]
looks great to me. my only comment is about introducing the word “demo” - usually we use “mock” or “mocked”.

**Comment by faboweb** [comment]
Do users understand the term "mocked"? This is why I used another terminology.

**Comment by faboweb** [comment]
I also got rid of mocked mode flags as well. Voyager will always prompt now to switch to mocked mode which is more usable should avoid confusion.

**Comment by faboweb** [comment]
Motherf***ing big PR. I apologize to reviewers in advance.

**Comment by NodeGuy** [comment]
> Do users understand the term "mocked"? This is why I used another terminology.

No, they don't.  "demo" is better for users.

**Comment by jolesbi** [comment]
the question of demo or mocked is a good one. we tend to assume more technical knowledge with our users than in a conventional consumer product by requiring keys, showing IP addresses, and referring to testnets. so mocked fits in with this technical user profile in my mind. though, demo would be ok too. the important thing for me is just consistency. 

i don't think the user should click the "try demo" button and then see the reference to a mocked server. consistency would be better here - no matter the choice of words.

**Comment by faboweb** [comment]
I think "mocked" is a term mostly familiar with software developers.
Technically interested people will still get "test network" and "ip
address". I propose we switch to "demo" network everywhere. What do you
think?

Jordan Bibla <notifications@github.com> schrieb am Di., 19. Juni 2018,
08:54:

**Comment by jolesbi** [comment]
demo sounds good to me (we discussed in the team meeting today)!

**Comment by nylira** [comment]
Let's clarify it further and rename it "offline-demo".

**Pull request merged**